sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuBroadcastToCpuExec.scala

```diff
@@ -0,0 +1,238 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.sql.rapids.execution
+
+import java.util.{Optional, UUID}
+import java.util.concurrent.{Callable, Future, TimeoutException, TimeUnit}
+
+import scala.collection.JavaConverters._
+import scala.collection.mutable.ListBuffer
+import scala.concurrent.Promise
+import scala.util.control.NonFatal
+
+import ai.rapids.cudf.{HostColumnVector, NvtxColor}
+import com.nvidia.spark.rapids.{GpuExec, GpuMetric, MetricRange, NvtxWithMetrics, RapidsHostColumnVector}
+import com.nvidia.spark.rapids.RapidsPluginImplicits.AutoCloseableProducingArray
+
+import org.apache.spark.SparkException
+import org.apache.spark.broadcast.Broadcast
+import org.apache.spark.launcher.SparkLauncher
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.{Attribute, SortOrder, UnsafeProjection, UnsafeRow}
+import org.apache.spark.sql.catalyst.plans.physical.{BroadcastMode, Partitioning}
+import org.apache.spark.sql.execution.{SparkPlan, SQLExecution, UnaryExecNode}
+import org.apache.spark.sql.execution.exchange.BroadcastExchangeExec.MAX_BROADCAST_TABLE_BYTES
+import org.apache.spark.sql.execution.metric.SQLMetrics
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.vectorized.ColumnarBatch
+import org.apache.spark.util.KnownSizeEstimation
+
+/**
+ * This is a specialized version of GpuColumnarToRow that wraps a GpuBroadcastExchange and
+ * converts the columnar results containing cuDF tables into Spark rows so that the results
+ * can feed a CPU BroadcastHashJoin. This is required for exchange reuse in AQE.
+ *
+ * @param child Input to broadcast
+ */
+case class GpuBroadcastToCpuExec(mode: BroadcastMode, child: SparkPlan)
```

Comment:
Can we have a follow on issue to look at how much of the code we can combine with `GpuBroadcastExchangeExec`. The reason I don't ask to do it here is because this is a P1 issue, and I don't want to get in the way of shipping.

Inside `GpuBroadcastExchangeExec` there are two APIs `doExecuteBroadcast` and `executeColumnarBroadcast`. Our code always calls the second one. The CPU code calls the first one but it is implemented to do exactly the same thing. At a minimum the first one should throw an exception but we probably could have it do the same thing that is happening here.  We might even be able to figure out a way so we broadcast the data twice, but we don't execute the broadcast twice.

Reply:
Here is the follow up issue to remove code duplication: https://github.com/NVIDIA/spark-rapids/issues/2219